expression_utils.go

```diff
@@ -0,0 +1,105 @@
+package gojsonsm
+
+import "fmt"
+
+func fieldExprMatches(lhs FieldExpr, rhs FieldExpr) bool {
+	if lhs.Root != rhs.Root {
+		return false
+	}
+	if len(lhs.Path) != len(rhs.Path) {
+		return false
+	}
+	for pathIdx := range lhs.Path {
+		if lhs.Path[pathIdx] != rhs.Path[pathIdx] {
+			return false
+		}
+	}
+	return true
+}
+
+func fetchExprFieldRefsRecurse(expr Expression, loopVars []VariableID, fields []FieldExpr) []FieldExpr {
+	switch expr := expr.(type) {
+	case FieldExpr:
+		isLoopVarRef := false
+		for _, loopVar := range loopVars {
+			if expr.Root == loopVar {
+				isLoopVarRef = true
+				break
+			}
+		}
+		if isLoopVarRef {
```

Comment:
Is it possible for a field var to be both a loop variable AND an actual field variable?

Reply:
It technically is, this particular logic is just about making sure that we don't include field references for loops that are inside the expression in the list of field references.  For instance, it makes sure that if we have an expression like `ANY $1 IN $doc.friends DO $1.id = $doc.bestFriendId END`, that the root expressions list will be `[$doc.friends, $doc.bestFriendId]` (aka: it excludes the reference to `$1.id`).  However if you were to gather the fieldExpr's for the loops body, then you would find `[$1.id, $doc.bestFriendId]`.  This particular ignorance of loop variables is integral to the new code which uses this information to calculate dependancies for determining the correct location to place after blocks.